command/login/aws.go

```diff
@@ -0,0 +1,138 @@
+	fs := flag.NewFlagSet("", flag.ContinueOnError)
+	fs.BoolVar(&a.autoBearerToken, "aws-auto-bearer-token", false,
+		"Automatically discover AWS credentials, construct the bearer token, and login. [aws-iam only]")
+
+	fs.BoolVar(&a.includeEntity, "aws-include-entity", false,
+		"Include a signed request to get the IAM role or IAM user in the bearer token. [aws-iam only]")
+
+	fs.StringVar(&a.stsEndpoint, "aws-sts-endpoint", "",
+		"URL for AWS STS API calls. [aws-iam only]")
+
+	fs.StringVar(&a.region, "aws-region", "",
+		"Region for AWS STS API calls. If set, should match the region of -aws-sts-endpoint. [aws-iam only]")
+
+	fs.StringVar(&a.serverIDHeaderValue, "aws-server-id-header-value", "",
+		"If set, an X-Consul-IAM-ServerID header is included in signed AWS API request(s) that form "+
+			"the bearer token. This value must match the server-side configured value for the auth method "+
+			"in order to login. This is optional and helps protect against replay attacks. [aws-iam only]")
+
+	fs.StringVar(&a.accessKeyId, "aws-access-key-id", "",
+		"AWS access key id to use. Requires -aws-secret-access-key if specified. [aws-iam only]")
+
+	fs.StringVar(&a.secretAccessKey, "aws-secret-access-key", "",
+		"AWS secret access key to use. Requires -aws-access-key-id if specified. [aws-iam only]")
+
+	fs.StringVar(&a.sessionToken, "aws-session-token", "",
+		"AWS session token to use. Requires -aws-access-key-id and -aws-secret-access-key if "+
+			"specified. [aws-iam only]")
+	return fs
+}
+
+// checkFlags validates flags for the aws-iam auth method.
+func (a *AWSLogin) checkFlags() error {
+	if !a.autoBearerToken {
+		if a.includeEntity || a.stsEndpoint != "" || a.region != "" || a.serverIDHeaderValue != "" ||
+			a.accessKeyId != "" || a.secretAccessKey != "" || a.sessionToken != "" {
+			return fmt.Errorf("Missing '-aws-auto-bearer-token' flag")
+		}
+	}
+	if a.accessKeyId != "" && a.secretAccessKey == "" {
+		return fmt.Errorf("Missing '-aws-secret-access-key' flag")
+	}
+	if a.secretAccessKey != "" && a.accessKeyId == "" {
+		return fmt.Errorf("Missing '-aws-access-key-id' flag")
+	}
+	if a.sessionToken != "" && (a.accessKeyId == "" || a.secretAccessKey == "") {
+		return fmt.Errorf("Missing '-aws-access-key-id' and '-aws-secret-access-key' flags")
+	}
+	return nil
+}
+
+// createAWSBearerToken generates a bearer token string for the AWS IAM auth method.
+// It will discover AWS credentials which are used to sign AWS API requests.
+// Alternatively, static credentials can be passed as flags.
+//
+// The bearer token contains a signed sts:GetCallerIdentity request.
+// If aws-include-entity is specified, a signed iam:GetRole or iam:GetUser request is
+// also included. The AWS credentials are used to retrieve the current user's role
+// or user name for the iam:GetRole or iam:GetUser request.
+func (a *AWSLogin) createAWSBearerToken() (string, error) {
+	cfg := aws.Config{
+		Endpoint: aws.String(a.stsEndpoint),
```

Comment:
Are `stsEndpoint` and `region` required? I don't see verification implying they are .

Reply:
Agree, needs clarification. They are not required CLI flags.

The AWS SDK needs a region, but there are various standard ways to pass the region (and other config) to the SDK aside from the flags we support here ([env vars, file, and others](https://docs.aws.amazon.com/sdk-for-go/api/aws/session/)).

Reply:
I clarified this a bit in the flag descriptions. I also fixed the logic here so that this always checks that the region and credentials were either passed or found by the SDK's credential chain.

(fwiw, the login attempt will fail if creds or region are missing, but the error message is more obtuse in that case)